src/Libraries/Microsoft.Extensions.AI/ChatCompletion/AnonymousDelegatingChatClient.cs

```diff
@@ -0,0 +1,213 @@
+
+        _completeFunc = completeFunc;
+        _completeStreamingFunc = completeStreamingFunc;
+    }
+
+    /// <inheritdoc/>
+    public override Task<ChatCompletion> CompleteAsync(
+        IList<ChatMessage> chatMessages, ChatOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(chatMessages);
+
+        if (_sharedFunc is not null)
+        {
+            return CompleteViaSharedAsync(chatMessages, options, cancellationToken);
+
+            async Task<ChatCompletion> CompleteViaSharedAsync(IList<ChatMessage> chatMessages, ChatOptions? options, CancellationToken cancellationToken)
+            {
+                ChatCompletion? completion = null;
+                await _sharedFunc(chatMessages, options, async (chatMessages, options, cancellationToken) =>
+                {
+                    completion = await InnerClient.CompleteAsync(chatMessages, options, cancellationToken).ConfigureAwait(false);
+                }, cancellationToken).ConfigureAwait(false);
+
+                if (completion is null)
+                {
+                    throw new InvalidOperationException("The wrapper completed successfully without producing a ChatCompletion.");
+                }
+
+                return completion;
+            }
+        }
+        else if (_completeFunc is not null)
+        {
+            return _completeFunc(chatMessages, options, InnerClient, cancellationToken);
+        }
+        else
+        {
+            Debug.Assert(_completeStreamingFunc is not null, "Expected non-null streaming delegate.");
+            return _completeStreamingFunc!(chatMessages, options, InnerClient, cancellationToken)
+                .ToChatCompletionAsync(coalesceContent: true, cancellationToken);
+        }
+    }
+
+    /// <inheritdoc/>
+    public override IAsyncEnumerable<StreamingChatCompletionUpdate> CompleteStreamingAsync(
+        IList<ChatMessage> chatMessages, ChatOptions? options = null, CancellationToken cancellationToken = default)
+    {
+        _ = Throw.IfNull(chatMessages);
+
+        if (_sharedFunc is not null)
+        {
+            var updates = Channel.CreateBounded<StreamingChatCompletionUpdate>(1);
+
+#pragma warning disable CA2016 // explicitly not forwarding the cancellation token, as we need to ensure the channel is always completed
+            _ = Task.Run(async () =>
+#pragma warning restore CA2016
+            {
+                Exception? error = null;
+                try
+                {
+                    await _sharedFunc(chatMessages, options, async (chatMessages, options, cancellationToken) =>
+                    {
+                        await foreach (var update in InnerClient.CompleteStreamingAsync(chatMessages, options, cancellationToken).ConfigureAwait(false))
+                        {
+                            await updates.Writer.WriteAsync(update, cancellationToken).ConfigureAwait(false);
```

Comment:
That's pretty clever. Do you think this unifying pattern can be extended beyond delegating wrappers? IOW could the behaviour of chat client be fully defined in terms of a variation of this higher-order function, and we expose `CompleteAsync` and `CompleteAsync` as extension methods?

I would guess not since the nested callback inherently relies on an inner client with discrete streaming and non-streaming functionalities.

Reply:
Yeah, I don't see a good way to do that. It is possible to define non streaming in terms of streaming, though, albeit with some minimal loss of fidelity.